lib/cdo/firehose.rb

```diff
@@ -123,6 +129,12 @@ def put_record(stream, data)
     raise ArgumentError.new("stream must be defined") if stream.nil? || stream.blank?
     raise ArgumentError.new("Stream #{stream} not found in STREAMS") if (stream_name = STREAMS[stream]).nil?
 
+    if data[:data_json].to_s.bytesize > MAX_DATA_JSON_BYTES
+      raise ArgumentError.new("data_json column too large (#{data[:data_json].to_s.bytesize} bytes)")
+    end
+    if data[:data_string].to_s.bytesize > MAX_DATA_STRING_BYTES
+      raise ArgumentError.new("data_string column too large (#{data[:data_string].to_s.bytesize} bytes)")
+    end
     record = add_common_values(data).to_json
     if (size = record.bytesize) > BYTES_PER_RECORD
       raise ArgumentError.new("Record too large (#{size} bytes)")
```

Comment:
Thanks for adding the new error catches Dave! Can you help me understand how data that that exceeds the max exits this function and still make it into firehose? I've read through the file but can't tell from a first read.

Reply:
data that exceeds the max does exit this function, but will not still make it into firehose. does that help?

Reply:
It’s pretty convoluted and it was mostly configured manually in AWS, so there isn’t much code that you can search for showing how it works. Here’s my memory of the architecture:

1. Various components in our Ruby web application servers invoke this library (`firehouse.rb`) to publish an Event 
2. This library publishes an Event in an AWS Firehose Stream
3. The AWS Firehose stream writes events as JSON text in batch text files in AWS S3 Objects (each file contains a long list of JSON objects).
4. A different AWS Firehose stream asynchronously picks ups those S3 Objects and executes a SQL statement on our Redshift cluster to INSERT each of the JSON objects into the Redshift `analysis.events` table, mapping specific properties in the JSON object to columns in the `events` table.
5. If a single JSON event within a batch file contains a value in one of its attributes that violates the data integrity rules (length, datatype, etc.) of the corresponding column in the `analysis.events` table THE ENTIRE BATCH OF EVENTS fails to load into Redshift. This means that a single event that has a `data_json` or `data_string` attribute that’s too long, can cause the other 1-2 thousand events in that batch to fail to load.

Reply:
Okay this is helpful- thank you both! I now realize that we are sacrificing a small amount of data so it doesn't fail the entire batch. So firehose will be *slightly undercounting events, but on a tiny scale?

Reply:
> So firehose will be *slightly undercounting events, but on a tiny scale?

yes, but only for firehose events which break the rules, and we'll now get notified about these cases so we can fix them. we should assume that the vast majority of firehose event types are written in a way that it'll never break these rules.